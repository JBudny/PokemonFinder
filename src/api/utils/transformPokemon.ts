import { Pokemon } from '@api/types/Pokemon'

import { PokemonItem } from '../types/pokemonItem.types'

interface TransformedPokemon {
	data: PokemonItem
}

export const transformPokemon = (pokemon: Pokemon): TransformedPokemon => {
	const { abilities, id, name, sprites, stats } = pokemon

	return {
		data: {
			abilities: abilities
				? abilities.map((ability) => ability.ability.name).join(', ')
				: '',
			id: id.toString(),
			name,
			stats,
			sprite: sprites.other['official-artwork'].front_default
		}
	}
}
