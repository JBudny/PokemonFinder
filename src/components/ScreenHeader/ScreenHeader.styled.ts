import { View } from 'react-native'

import styled from 'styled-components/native'

export const StyledScreenHeader = styled(View)`
	background-color: ${({ theme }) => theme.palette.primary};
	padding: ${({ theme }) => theme.designSystem.spacingLarge}px;
	height: 180px;
`
