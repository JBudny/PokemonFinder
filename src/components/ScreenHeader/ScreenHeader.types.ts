export interface ScreenHeaderProps {
	text: string
}
