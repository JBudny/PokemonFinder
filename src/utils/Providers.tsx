import React, { FC } from 'react'
import { Provider } from 'react-redux'

import { ThemeProvider } from 'styled-components/native'

import { AppStore, setupStore } from '@src/store'
import { schema } from '@theme/schema'

const originalStore = setupStore()

interface ProvidersProps {
	store: AppStore
}

export const ProvideTheme: FC = ({ children }) => (
	<ThemeProvider theme={schema.light}>{children}</ThemeProvider>
)

export const Providers: FC<ProvidersProps> = ({
	children,
	store = originalStore
}) => (
	<Provider store={store}>
		<ProvideTheme>{children}</ProvideTheme>
	</Provider>
)
