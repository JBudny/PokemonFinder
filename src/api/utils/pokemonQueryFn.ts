import { BaseQueryApi } from '@reduxjs/toolkit/dist/query/baseQueryTypes'

import { transformPokemon } from './transformPokemon'

/**
 * Custom query function for the Pokemon API with error handling behavior
 * and data transformation logic.
 */
export const pokemonQueryFn = async (
	nameArg: string,
	_queryApi: BaseQueryApi,
	_extraOptions: Record<string, unknown>,
	fetchWithBQ: any
) => {
	const { data } = await fetchWithBQ(`/pokemon/${nameArg.toLowerCase()}`)

	try {
		return transformPokemon(data)
	} catch (e) {
		return {
			error: {
				status: 404,
				statusText: 'Pokemon not found!',
				data: undefined
			}
		}
	}
}
