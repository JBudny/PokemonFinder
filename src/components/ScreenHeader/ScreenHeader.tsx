import React, { FC } from 'react'

import { Heading } from '../Typography/Heading/Heading'
import { StyledScreenHeader } from './ScreenHeader.styled'
import { ScreenHeaderProps } from './ScreenHeader.types'

/**
 * Screen header.
 * @param text
 */
export const ScreenHeader: FC<ScreenHeaderProps> = ({ text }) => (
	<StyledScreenHeader>
		<Heading text={text} size="large" />
	</StyledScreenHeader>
)
