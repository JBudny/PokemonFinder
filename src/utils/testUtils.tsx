import React from 'react'

import type { PreloadedState } from '@reduxjs/toolkit'
import type { RenderOptions } from '@testing-library/react-native'
import { render } from '@testing-library/react-native'

import type { AppStore, RootState } from '@src/store'
import { setupStore } from '@src/store'

import { Providers } from './Providers'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store. For
// future dependencies, such as wanting to test with react-router, you can extend
// this interface to accept a path and route and use those in a <MemoryRouter />
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
	preloadedState?: PreloadedState<RootState>
	store?: AppStore
}

const renderWithProviders = (
	ui: React.ReactElement,
	{
		preloadedState = {},
		store = setupStore(preloadedState),
		...renderOptions
	}: ExtendedRenderOptions = {}
) => {
	const Wrapper: React.FC = ({ children }) => (
		<Providers store={store}>{children}</Providers>
	)

	return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

export { renderWithProviders }
