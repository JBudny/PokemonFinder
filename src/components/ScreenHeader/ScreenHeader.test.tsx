import React from 'react'

import { renderWithProviders } from '@src/utils/testUtils'

import { ScreenHeader } from './ScreenHeader'

test('ScreenHeader component should display screen title', () => {
	const { getByText } = renderWithProviders(<ScreenHeader text="Test title" />)

	expect(getByText('Test title')).toBeDefined()
})

test('ScreenHeader component should match the snapshot', () => {
	const screenHeader = renderWithProviders(<ScreenHeader text="Test title" />)

	expect(screenHeader).toMatchSnapshot()
})
