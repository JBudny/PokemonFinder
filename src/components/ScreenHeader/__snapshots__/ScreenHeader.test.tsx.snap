// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScreenHeader component should match the snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "rgb(39, 93, 173)",
        "height": 180,
        "paddingBottom": 15,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 15,
      },
    ]
  }
>
  <Text
    size="large"
    style={
      Array [
        Object {
          "color": "rgb(255, 255, 255)",
          "fontSize": 35,
          "fontWeight": "bold",
        },
      ]
    }
  >
    Test title
  </Text>
</View>
`;
